### PR TITLE
calendar-rework

### PR DESCRIPTION
**Pull request by fdamhaut**
Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/calendar_bis/models/calendar_event.py [hunk 1]
```diff
@@ -107,6 +110,23 @@ def merge_timeslots(self, values, ts_to_skip=None):
...
+
+    def break_after(self, slot):
+        self.ensure_one()
+        post_break = self.timeslot_ids.filtered(lambda x: x.start >= slot.start)
+        if self.count:
+            post_break.event_id = self.copy({'count': len(post_break)})
+            self.count -= len(post_break)
+        else:
+            post_break.event_id = self.copy()
+            self.until = slot.start - relativedelta(days=1)
```
Comment (leonardopavanrocha) [inline addons/calendar_bis/models/calendar_event.py:127]: need to check if there are no edge cases here :thinking:

File: addons/calendar_bis/models/calendar_event.py [hunk 2]
```diff
@@ -107,6 +110,23 @@ def merge_timeslots(self, values, ts_to_skip=None):
         (to_unlink + current).unlink()
         return to_create + values
 
+    def exdate(self, timeslot):
+        self.ensure_one()
+        self.exdated = self.exdated + ',' + str(timeslot.start.date()) if self.exdated else str(timeslot.start.date())
```
Comment (leonardopavanrocha) [inline addons/calendar_bis/models/calendar_event.py:115]: I guess we don't need conditions here

File: addons/calendar_bis/models/calendar_event.py [hunk 3]
```diff
@@ -82,10 +81,31 @@ def make_timeslots(self, current_ts=None):
...
+            if not cmp: # Same start date
+                current[0].write(values[0])
+                current, values = current[1:], values[1:]
+            elif cmp < 0: # Current is before values
+                to_unlink += current[0]
+                current = current[1:]
+            else :  # Current is after values
+                to_create.append(values[0])
+                values = values[1:]
+        (to_unlink + current).unlink()
```
[the remark this answers is not in the document]
Comment (fdamhaut) [reply addons/calendar_bis/models/calendar_event.py]: The problem with your version is that in the else case, current should not be be "advanced'  but it is.
That being said, I could use generators instead of list to avoid recreating everytime

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 4]
```diff
@@ -69,6 +70,24 @@ def _compute_access(self):
...
+    # DRAG AND DROP
+    def drag_timeslot(self, dt):
+        if not self.is_recurring:
+            self.start = dt
+        elif self.freq in ['daily', 'yearly']:
+            self.start = dt
+            self.event.make_timeslot(self)
+        elif self.freq == 'weekly':
+            self.setattr(days[self.start.weekday], True)
+            self.setattr(days[dt.weekday], True)
```
[the remark this answers is not in the document]
Comment (fdamhaut) [reply addons/calendar_bis/models/calendar_timeslot.py:110]: To avoid having a 7way elif for "translating" day number to field

File: addons/calendar_bis/models/calendar_event.py (hunk 2 above)
Comment (fdamhaut) [reply addons/calendar_bis/models/calendar_event.py:115]: default is '', so if we exdate ad date, we end up with `,<date>` and we might as well have nice strings

File: addons/calendar_bis/models/calendar_event.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (leonardopavanrocha) [reply addons/calendar_bis/models/calendar_event.py]: Ah, I missed that indeed

File: addons/calendar_bis/models/calendar_timeslot.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (leonardopavanrocha) [reply addons/calendar_bis/models/calendar_timeslot.py:110]: doing `self[days[self.start.weekday]] = True` doesn't work?

File: addons/calendar_bis/models/calendar_event.py (hunk 2 above)
Comment (leonardopavanrocha) [reply addons/calendar_bis/models/calendar_event.py:115]: Ah, I was getting the wrong order. I assumed it was `self.exdated + ',' + (x if self.exdated else x)` :facepalm:

File: addons/calendar_bis/models/calendar_event.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    timeslot_ids = fields.One2many('calendar.timeslot', 'event_id')
+
+    is_public = fields.Boolean(default=False, public=True)
+    partner_id = fields.Many2one('res.partner', string="Owner", public=True, default=lambda self: self.env.user.partner_id.id)
+
+    name = fields.Char(public_default="Busy")
+    note = fields.Char()
+
+    # Recurrence
+    is_recurring = fields.Boolean(default=False, public=True)
```
Comment (arj-odoo) [inline addons/calendar_bis/models/calendar_event.py:28]: for now, the recurrence is handled by several fields:
recurrency: bool to know if the event is in a recurrence
recurrence_id: the linked recurrence
follow_recurrence: to take into account events that does not follow the rule. You use exdated for that purpose I think. But you don't have an object to store the recurrence. I think your model is closer to what microsoft or google api are sending.

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+    can_read_private = fields.Boolean(compute='_compute_access', default=True)
+    can_write = fields.Boolean(compute='_compute_access', default=True)
+    active = fields.Boolean(default=True)
+    edit = fields.Selection([('one', 'This event only'),
+                             ('all', 'All events in the series'),
+                             ('post', 'All event after this one')], default='all', store=False)
+
+    # Time Related Fields
+    start = fields.Datetime(default=fields.Datetime.now, required=True)
+    stop = fields.Datetime(required=True, compute='_compute_stop', readonly=False, store=True)
```
Comment (arj-odoo) [inline addons/calendar_bis/models/calendar_timeslot.py:24]: In the current implementation, we have start, stop and start_date and stop_date it allows to display all_day events. I see that you removed these fields. In calendar we have two complex cases to handle:

1. Jerémy needs to join a meeting online at 17:00. But the organizer, Alan is in another timezone. The system will display 17:30 to Jeremy and it will display 12:30 to Alan. I think your implementation should work out of the box for that case. 
2. The other case which is more difficult is the event common for several people but occurring at the same date. Alain and Jeremy will be off for Christmas day but it should display the whole day for both of them.
3. Moreover, their company is giving them the 24th of December off. It should be displayed for all of them at the same time. 

The second situation is quite difficult to handle. We somewhat support it with the all day events that have a forced date and the _inverse_dates method that will force the datetime of all day events from 8 to 18. It works only if you are not too far in Japan or extreme Russia. How to you plan to handle these kind of things?

**Review by arj-odoo — COMMENTED** [review]
Congratulations :-). That's a huge amount of work. 
I have a few questions about the all days events and a few remarks. We can discuss them face to face at the office. Thanks :-).

File: addons/calendar_bis/util/RRULE.py [hunk 7]
```diff
@@ -0,0 +1,131 @@
...
+    ('mo', rrule.MO),
+    ('tu', rrule.TU),
+    ('we', rrule.WE),
+    ('th', rrule.TH),
+    ('fr', rrule.FR),
+    ('sa', rrule.SA),
+    ('su', rrule.SU),
+]
+wd2str = {x[1].weekday:x[0] for x in RRULE_WEEKDAYS}
+str2day = {x[0]:x[1] for x in RRULE_WEEKDAYS}
```
[the remark this answers is not in the document]
Comment (fdamhaut) [reply addons/calendar_bis/util/RRULE.py:27]: It's just to translate day number to day `rrule` id (first 2 letters of the day) it will never be translated

File: addons/calendar_bis/models/calendar_timeslot.py (hunk 6 above)
Comment (fdamhaut) [reply addons/calendar_bis/models/calendar_timeslot.py]: Yes, All day is not implemented yet but will be in the final version

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 8]
```diff
@@ -0,0 +1,234 @@
...
+            attendee_mails = slot.attendee_ids.mapped('email')
+            slot.attendee_ids += slot.attendee_ids.create(
+                [{'partner_id': pid.id, 'timeslot_id': slot.id} for pid in slot.partner_ids if pid not in attendee_pids] + \
+                [{'email': attmail.name, 'timeslot_id': slot.id} for attmail in slot.attendee_mail_ids if attmail not in attendee_mails])
+        return res
+
+    # ACCESS FUNCTIONS
+    def _has_access(self, access):
+        self.ensure_one()
+        if isinstance(self.id, models.NewId): return True
```
Comment (tivisse) [inline addons/calendar_bis/models/calendar_timeslot.py:112]: Check that existing records in the context of a compute won't be available to the user.

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 9]
```diff
@@ -0,0 +1,234 @@
...
+    def copy_data(self, default=None):
+        data = super().copy_data(default)[0]
+        return [{**data, 'attendee_ids': [Command.create(attendee.copy_data()[0]) for attendee in self.attendee_ids]}]
+
+    #### RELATED ####
+    @api.depends('event_id.name')
+    @api.depends_context('uid')
+    def _compute_name(self):
+        for slot in self:
+            slot.name = slot.event_id.name if slot.can_read_private else 'Busy'
```
Comment (tivisse) [inline addons/calendar_bis/models/calendar_timeslot.py:212]:
```suggestion
            slot.name = slot.event_id.name if slot.can_read_private else _('Busy')
```

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 10]
```diff
@@ -0,0 +1,234 @@
...
+                            string='Frequency', default='weekly', compute="_compute_recurring", inverse="_inverse_recurring")
+    until = fields.Datetime(compute="_compute_recurring", inverse="_inverse_recurring")
+    count = fields.Integer(compute="_compute_recurring", inverse="_inverse_recurring")
+    interval = fields.Integer(compute="_compute_recurring", inverse="_inverse_recurring")
+    monthday = fields.Integer(compute="_compute_recurring", inverse="_inverse_recurring")
+    monthweekday_n = fields.Integer(compute="_compute_recurring", inverse="_inverse_recurring")
+    monthweekday_day = fields.Selection([('mo', 'Monday'), ('tu', 'Tuesday'), ('we', 'Wednesday'),
+        ('th', 'Thursday'), ('fr', 'Friday'), ('sa', 'Saturday'), ('su', 'Sunday')], compute="_compute_recurring", inverse="_inverse_recurring")
+
+    def write(self, values):
```
Comment (tivisse) [inline addons/calendar_bis/models/calendar_timeslot.py:61]: Add comment pliz

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 11]
```diff
@@ -0,0 +1,234 @@
...
+        for slot, vals in zip(res, values):
+            if not 'partner_ids' in vals and not 'attendee_mail_ids' in vals:
+                continue
+            slot.partner_ids = vals.get('partner_ids', [])
+            slot.attendee_mail_ids = vals.get('attendee_mail_ids', [])
+            attendee_pids = slot.attendee_ids.partner_id.ids
+            attendee_mails = slot.attendee_ids.mapped('email')
+            slot.attendee_ids += slot.attendee_ids.create(
+                [{'partner_id': pid.id, 'timeslot_id': slot.id} for pid in slot.partner_ids if pid not in attendee_pids] + \
+                [{'email': attmail.name, 'timeslot_id': slot.id} for attmail in slot.attendee_mail_ids if attmail not in attendee_mails])
```
Comment (tivisse) [inline addons/calendar_bis/models/calendar_timeslot.py:106]: Should be validated (do we really want to do this), and probably moved into an independent PR.

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 12]
```diff
@@ -0,0 +1,234 @@
...
+            slot.duration = (slot.stop - slot.start).total_seconds() / 3600
+
+    def copy(self, values):
+        if 'attendee_ids' not in values:
+            values['attendee_ids'] = [Command.create(attendee.copy_data()[0]) for attendee in self.attendee_ids]
+        return super().copy(values)
+
+    def copy_data(self, default=None):
+        data = super().copy_data(default)[0]
+        return [{**data, 'attendee_ids': [Command.create(attendee.copy_data()[0]) for attendee in self.attendee_ids]}]
```
Comment (tivisse) [inline addons/calendar_bis/models/calendar_timeslot.py:205]: copy=True on attendee_ids should do the trick + copy=False on attendee fields you don't want to copy (state, ...)

File: addons/calendar_bis/models/calendar_timeslot.py [hunk 13]
```diff
@@ -0,0 +1,235 @@
...
+        for slot in self:
+            slot.event_id.note = slot.note
+
+    def mass_delete(self, update_policy):
+        self.ensure_one()
+        if not self.can_write:
+            raise AccessError(_("You don't have access to delete this timeslot."))
+        if update_policy == "all_events":
+            return self.event_id.unlink()
+        self.event_id.break_after(self).unlink()
```
Comment (arj-odoo) [inline addons/calendar_bis/models/calendar_timeslot.py:235]: :eyes:

File: addons/calendar_bis/security/ir.model.access.csv [hunk 14]
```diff
@@ -0,0 +1,5 @@
+id,name,model_id:id,group_id:id,perm_read,perm_write,perm_create,perm_unlink
+calendar_bis.access_calendar_event,access_calendar_event,calendar_bis.model_calendar_event,base.group_user,1,1,1,1
+calendar_bis.access_calendar_timeslot,access_calendar_timeslot,calendar_bis.model_calendar_timeslot,base.group_user,1,1,1,1
+calendar_bis.access_calendar_attendee,access_calendar_attendee,calendar_bis.model_calendar_attendee,base.group_user,1,1,1,1
+calendar_bis.access_calendar_attendee_mail,access_calendar_attendee_mail,calendar_bis.model_calendar_attendee_mail,base.group_user,1,1,1,1
```
Comment (arj-odoo) [inline addons/calendar_bis/security/ir.model.access.csv:5]: :eyes:
